### PR TITLE
tests: parallel images build support

### PR DESCRIPTION
**Pull request by marcov**
Rework test_image.sh and Makefile to allow building images in parallel
for faster test execution.
Add some new targets to Makefile ({rootfs,image,initrd}-all,
list-distros).

Fixes: #168

### CODE REVIEW

File: tests/test_images.sh [hunk 1]
```diff
@@ -318,230 +274,153 @@ install_initrd_create_container()
...
 
-		local rootfs_size=$(du -sb "${rootfs}" | awk '{print $1}')
+	echo "will build with tgts: ${makeTargets[@]}"
+	echo "will build with vars: ${makeVars[@]}"
 
-		if [ "$type" = "image" ]
-		then
-			# Images need systemd
-			[ "$opt" = "init" ] && continue
+	sudo -E make -j ${makeTargets[@]} ${makeVars[@]}
```
[the remark this answers is not in the document]
Comment (grahamwhaley) [reply tests/test_images.sh]: +1 - we use nproc in other places as well etc.

[the remark this answers is not in the document]
Comment (marcov) [reply tests/test_images.sh]: Actually `make` is run by two bash processes in background (to build in parallel with and without `AGENT_INIT=yes`). So, to play it safe, should it be `$(nproc) / 2 -1` ?

```
 4220 pts/2    S+     0:00  \_ /bin/bash ./tests/test_images.sh
 4455 pts/2    S+     0:00      \_ /bin/bash ./tests/test_images.sh
 4460 pts/2    S+     0:00      |   \_ sudo -E make -j image-fedora image-centos image-ubuntu image-clearlinux image-eu
 4462 pts/2    S+     0:00      |       \_ make -j image-fedora image-centos image-ubuntu image-clearlinux image-eulero
 4510 pts/2    S+     0:00      |           \_ /bin/bash /home/marco/go/src/github.com/kata-containers/osbuilder/rootfs
 4656 pts/2    Sl+    0:00      |           |   \_ docker run --env https_proxy= --env http_proxy= --env AGENT_VERSION=
 4511 pts/2    S+     0:00      |           \_ /bin/bash /home/marco/go/src/github.com/kata-containers/osbuilder/rootfs
 4683 pts/2    Sl+    0:00      |           |   \_ docker run --env https_proxy= --env http_proxy= --env AGENT_VERSION=
 4515 pts/2    S+     0:00      |           \_ /bin/bash /home/marco/go/src/github.com/kata-containers/osbuilder/rootfs
 4634 pts/2    Sl+    0:00      |           |   \_ docker run --env https_proxy= --env http_proxy= --env AGENT_VERSION=
 4527 pts/2    S+     0:00      |           \_ /bin/bash /home/marco/go/src/github.com/kata-containers/osbuilder/rootfs
 4559 pts/2    S+     0:00      |           |   \_ /bin/bash /home/marco/go/src/github.com/kata-containers/osbuilder/ro
 4561 pts/2    S+     0:00      |           |       \_ curl -sL https://download.clearlinux.org/latest
 4529 pts/2    S+     0:00      |           \_ /bin/bash /home/marco/go/src/github.com/kata-containers/osbuilder/rootfs
 4671 pts/2    Sl+    0:00      |               \_ docker run --env https_proxy= --env http_proxy= --env AGENT_VERSION=
 4456 pts/2    S+     0:00      \_ /bin/bash ./tests/test_images.sh
 4459 pts/2    S+     0:00          \_ sudo -E make -j initrd-alpine USE_DOCKER=true ROOTFS_BUILD_DEST=/tmp/osbuilder-t
 4461 pts/2    S+     0:00              \_ make -j initrd-alpine USE_DOCKER=true ROOTFS_BUILD_DEST=/tmp/osbuilder-test.
 4509 pts/2    S+     0:00                  \_ /bin/bash /home/marco/go/src/github.com/kata-containers/osbuilder/rootfs
 4680 pts/2    Sl+    0:00                      \_ docker run --env https_proxy= --env http_proxy= --env AGENT_VERSION=
```

File: tests/test_images.sh [hunk 2]
```diff
@@ -318,230 +274,153 @@ install_initrd_create_container()
...
+}
 
-		build_rootfs "${distro}" "${rootfs}"
+call_make() {
+	targetType=$1
+	shift
+	makeVars=()
+	makeTargets=()
+	for t in $@; do
+		if [[ "$t" =~ .+= ]]; then
```
[the remark this answers is not in the document]
Comment (marcov) [reply tests/test_images.sh]: I'll reformat a bit to make it easier to read

**Comment by marcov** [comment]
I think the 2 build configurations in  `.travis.yml` with and without AGENT_INIT can be dropped now, as those are already built in parallel in `tests_images.sh`.

Do you agree?

PS: looking at the current code on master and at the Travis build logs, I am not sure this split is actually done.

**Comment by jodh-intel** [comment]
> I think the 2 build configurations in .travis.yml with and without AGENT_INIT can be dropped now, as those are already built in parallel in tests_images.sh.

Sounds good to me - is that ok with you @bergwolf?

File: tests/test_images.sh (hunk 1 above)
[the remark this answers is not in the document]
Comment (jodh-intel) [reply tests/test_images.sh]: Maybe this won't be so bad as I suspect most of the test is going to be network-bound but we just need to ensure we don't kill a developers box by running this test ;)

[the remark this answers is not in the document]
Comment (grahamwhaley) [reply tests/test_images.sh]: good thought @marcov Let's try nproc/2. Although... I have no idea how many procs our cloud test instances generally have - I can't see that info in a console log either. @chavafg - any ideas on that?
We might want to do some experimentation and monitor `top` before we decide on a setting - if we are net bound like @jodh-intel says, then we can bump it up.

[the remark this answers is not in the document]
Comment (marcov) [reply tests/test_images.sh]: We should also add the `CI` variable definition to `.travis.yml` to take full advantage of the parallel building.

**Comment by marcov** [comment]
Hey, could any of the approvers kindly kick the CI ? :stuck_out_tongue:

**Comment by grahamwhaley** [comment]
/test
was that the phrase...

**Comment by marcov** [comment]
I added `set -u`  to `test_images.sh` and the failure on `jenkins-ci-centos-7` is definitely a bug in my code (`$CI` should be `${CI:-}`).
```
/tmp/jenkins/workspace/kata-containers-osbuilder-centos-7-4-PR/go/src/github.com/kata-containers/osbuilder/.ci/../tests/test_images.sh: line 294: CI: unbound variable
/tmp/jenkins/workspace/kata-containers-osbuilder-centos-7-4-PR/go/src/github.com/kata-containers/osbuilder/.ci/../tests/test_images.sh: line 294: CI: unbound variable
ERROR: Background build job failed
INFO: ERROR: test failed
```
However I am kind of surprised that `CI` **is not** defined on `jenkins-ci-centos-7` ..!
@chavafg  can you verify if it is set in Jenkins config?

**Comment by chavafg** [comment]
we are getting this error, when running the tests:
```
docker: Error response from daemon: OCI runtime create failed: Failed to check if grpc server is working: rpc error: code = DeadlineExceeded desc = context deadline exceeded: unknown.
```

which I have seen lately in some bug reports from users.

**Comment by chavafg** [comment]
and found 
```
/tmp/jenkins/workspace/kata-containers-osbuilder-fedora-27-PR/go/src/github.com/kata-containers/osbuilder/.ci/../tests/test_images.sh: line 189: ID_LIKE: unbound variable
```

on the fedora job.

File: tests/test_images.sh [hunk 3]
```diff
@@ -185,7 +182,15 @@ set_runtime()
 	[ -z "$name" ] && die "need name"
 
 	# Travis doesn't support VT-x
-	[ -n "$TRAVIS" ] && return
+	[ -n "${TRAVIS:-}" ] && return
+
+	source /etc/os-release
+
+	if [[ "$ID_LIKE" =~ suse ]]; then
```
[the remark this answers is not in the document]
Comment (marcov) [reply tests/test_images.sh]: @chavafg @jodh-intel thanks.

About the `DeadlineExceeded`: is this related: kata-containers/runtime#702?

[the remark this answers is not in the document]
Comment (jodh-intel) [reply tests/test_images.sh]: Hi @marcov - yep, I think so.

**Comment by jodh-intel** [comment]
Hi @marcov - I spot a problem! If you look at the 16.04 log for example, and search for "INFO: osbuilder metadata file for", you'll see that *multiple* YAML files have been concatenated into `osbuilder.yaml`.

**Comment by marcov** [comment]
@jodh-intel: no, that's actually the output of multiple files, but the spacing make it look like it's a single file :). I'll try to make it  more readable.

```
touch /tmp/osbuilder-test.UsE7M2P/rootfs-osbuilder/.euleros_rootfs.done
INFO: osbuilder metadata file for fedora:

---
osbuilder:
  url: "https://github.com/kata-containers/osbuilder"
  version: "unknown"
rootfs-creation-time: "2018-09-21T16:30:20.420795130+0000Z"
...
  agent-is-init-daemon: "no"
INFO: osbuilder metadata file for centos:

---
osbuilder:
  url: "https://github.com/kata-containers/osbuilder"
  version: "unknown"
```

**Comment by jodh-intel** [comment]
Ah - thanks, that would make it clearer ;)

**Comment by marcov** [comment]
@jodh-intel: done.

Ok so now all failures are imputable to `DeadlineExceeded` (when testing locally all containers can run just fine).

**Comment by marcov** [comment]
Let's wait for `1.3.0` with urandom fix  to be available on OBS before retesting...

**Comment by jodh-intel** [comment]
@marcov - 1.3.0 is out now! ;)

**Comment by jodh-intel** [comment]
...errm, well yes it is, but not on OBS yet... :)

**Comment by marcov** [comment]
@jodh-intel, actually the `1.3.0` [is on OBS too.](http://download.opensuse.org/repositories/home:/katacontainers:/releases:/x86_64:/master/)

The problem is that the installation instructions still points to a legacy repo path, [e.g. for Ubuntu.](https://github.com/kata-containers/documentation/blob/master/install/ubuntu-installation-guide.md)

And [kata-manager](https://github.com/kata-containers/tests/tree/master/cmd/kata-manager) is using those instructions to install Kata on Jenkins nodes.

And maybe this is another thing to consider in kata-containers/documentation#234?

**Comment by jodh-intel** [comment]
@marcov - gah - I think I got browser-cached when checking OBS! :)

Yep - we need more input on that thread - thanks for yours btw!

**Comment by marcov** [comment]
PS: looks like including a comment with a URL containing  `/ t e s t` has as side effect triggering the CI :rofl:

**Comment by jodh-intel** [comment]
@chavafg - ^ :smile:

**Comment by grahamwhaley** [comment]
Heh - the CI trigger is done via a regexp. Maybe we need/are missing a `^` on the front of the pattern for such cases.

**Comment by marcov** [comment]
And a `$` too :)

**Comment by jodh-intel** [comment]
*... If I had a dollar for every caret...* :smile:

**Comment by chavafg** [comment]
ouch. maybe something like `.*^/test$.*` to support multilines? checking...

**Comment by marcov** [comment]
All green with the `urandom` fix. 
I am removing from `.travis.yml` the `AGENT_INIT=yes / no` combinations, since now the agent / systemd builds are run in parallel by the test script directly.

**Comment by jodh-intel** [comment]
@marcov - nice work! :)

**Comment by grahamwhaley** [comment]
Possible network related problem on 16.04 (or, could be our OBS). Giving it a nudge.
```
0 upgraded, 8 newly installed, 0 to remove and 81 not upgraded.
Need to get 60.8 MB of archives.
After this operation, 675 MB of additional disk space will be used.
Get:1 http://download.opensuse.org/repositories/home:/katacontainers:/releases:/x86_64:/master/xUbuntu_16.04  kata-containers-image 1.3.0-9 [30.1 MB]
Get:2 http://download.opensuse.org/repositories/home:/katacontainers:/releases:/x86_64:/master/xUbuntu_16.04  kata-ksm-throttler 1.3.0.git+6e903fb-11 [2,611 kB]
Get:3 http://download.opensuse.org/repositories/home:/katacontainers:/releases:/x86_64:/master/xUbuntu_16.04  kata-linux-container 4.14.67.12-10 [10.1 MB]
Get:4 http://download.opensuse.org/repositories/home:/katacontainers:/releases:/x86_64:/master/xUbuntu_16.04  kata-proxy 1.3.0+git.6ddb006-9 [724 kB]
Get:5 http://download.opensuse.org/repositories/home:/katacontainers:/releases:/x86_64:/master/xUbuntu_16.04  kata-shim 1.3.0+git.5fbf1f0-8 [2,297 kB]
Err:5 http://download.opensuse.org/repositories/home:/katacontainers:/releases:/x86_64:/master/xUbuntu_16.04  kata-shim 1.3.0+git.5fbf1f0-8
  Hash Sum mismatch
Get:6 http://download.opensuse.org/repositories/home:/katacontainers:/releases:/x86_64:/master/xUbuntu_16.04  qemu-lite 2.11.0+git.f886228056-12 [5,549 kB]
Get:7 http://download.opensuse.org/repositories/home:/katacontainers:/releases:/x86_64:/master/xUbuntu_16.04  qemu-vanilla 2.11.2+git.0982a56a55-12 [5,522 kB]
Get:8 http://download.opensuse.org/repositories/home:/katacontainers:/releases:/x86_64:/master/xUbuntu_16.04  kata-runtime 1.3.0+git.a786643-14 [3,887 kB]
Fetched 58.5 MB in 40s (1,457 kB/s)
INFO: ERROR: test failed

INFO: images:

total 0
INFO: rootfs:

ls: cannot access '/tmp/osbuilder-test.UpjOGiZ/rootfs-osbuilder': No such file or directory
Build step 'Execute shell' marked build as failure
```

**Comment by marcov** [comment]
one more ack to go :slightly_smiling_face:

**Pull request merged**